log.go

```diff
@@ -222,57 +217,38 @@ func CloseOutput() (err error) {
 	return std.CloseOutput()
 }
 
+func (log *Log) needsRotating() bool {
+	return log.logFileName != "" && log.currentByteCount > log.maxLogSize
+}
+
 func (log *Log) rotateLog() {
-	// No need to rotate a log file that doesn't exist...
-	if log.logFileName == "" {
-		return
-	}
+	// Close our current file
+	log.logWriter.Close()
 
-	// Acquire mutex to rotate
-	log.mu.Lock()
-	defer log.mu.Unlock()
+	// Delete any existing rotated file
+	rotateName := log.logFileName + ".prev"
+	os.Remove(rotateName)
 
-	fileInfo, err := os.Stat(log.logFileName)
+	// Copy the existing file over so we can truncate in place
+	dst, err := os.OpenFile(rotateName, os.O_WRONLY|os.O_CREATE, 0600)
 	if err != nil {
-		fmt.Printf("ERROR: Could not stat log file '%s' - err='%s'\n", log.logFileName, err.Error())
-
-		// Attempt recovery to prevent filling the filesystem...
-		log.closeOutput()
-		os.Remove(log.logFileName)
-		log.setOutput(log.logFileName)
-		return
-	}
-
-	// Rotate the log file if it grows too large
-	if fileInfo.Size() > log.maxLogSize {
-		// Close our current file
-		log.logWriter.Close()
-
-		// Delete any existing rotated file
-		rotateName := log.logFileName + ".prev"
-		os.Remove(rotateName)
-
-		// Copy the existing file over so we can truncate in place
-		dst, err := os.OpenFile(rotateName, os.O_WRONLY|os.O_CREATE, 0600)
+		fmt.Printf("ERROR: Could not open target file for rotation. err='%s'\n", err.Error())
+	} else {
+		src, err := os.OpenFile(log.logFileName, os.O_RDONLY, 0600)
```

Comment:
Yes it should be nil, but I think it's good practice to not shadow variables as it can lead to obscure issues.